### PR TITLE
Fixes #31485 - Unauthenticated repo cache blocks access to repositories that require authentication

### PR DESCRIPTION
**Pull request by ianballou**
# To test:

0) Set up a smart proxy with Pulp 3.  You can use the following installer options to install Pulp 3:
```
--foreman-proxy-plugin-pulp-pulpcore-enabled 1
--foreman-proxy-plugin-pulp-pulpcore-mirror 1
```
You might hit some bumps with the installation, so let me know if you're stuck.

1) Clone the smart-proxy repo and copy over all config settings from the system-installed foreman-proxy.  You'll have to copy the certs to vagrant's home folder so it can access them and update the config files accordingly.
2) Once your smart proxy is working, install this smart proxy plugin.  You can do this by cloning the repo, adding its gem to `smart-proxy/bundler.d/Gemfile.local./rb`, and then copying the sample config file from this smart proxy plugin to `smart-proxy/config/settings.d/container_gateway.yml`.
3) Create a database called `smart_proxy_container_gateway` in postgres.  You'll need to make sure the database user you set in the config can read and write to this database.
4) Sync a docker repo to your smart proxy
5) Look at your catalog:  ```sudo podman search `hostname`/```
    - This should be empty
6) Find the name of your docker repo from your Katello server.  It'll look like `default_organization-environment-content_view-product-image`
7) Try a pull that should fail "unauthorized": ```sudo podman pull `hostname`/<docker repo name>```
8) Add your docker repo to the unauthenticated repo list via `httpie`:
```
http PUT https://`hostname`/v2/unauthenticated_repository_list repositories:='["<docker repo name>"]'
```
9) View your catalog again. It should have the repo you added
10) Try to `podman pull` the docker repo again. It should work now.

The idea is that, at smart proxy sync time, Katello will send a list of unauthenticated repos to the smart proxy via that unauthenticated_repository_list endpoint.

# Alternative testing
Use [this Katello PR](https://github.com/Katello/katello/pull/9043) to automatically update the unauthenticated repo list at smart proxy sync time.

### CODE REVIEW

**Comment by jjeffers** [comment]
@ianballou meant to start testing this before the holiday last week.

I was wondering what if there are additional tests you were going to add to this repo?

**Comment by ianballou** [comment]
@jjeffers I need to setup a rubocop check, redmine check, and tests at some point.  I'm thinking I'll do that all together either before or during my next feature PR (which will complete the initial functionality of this smart proxy plugin).

**Comment by jjeffers** [comment]
@ianballou some questions, after setting up the smart proxy and enabling the plugin:

1. should we see the smart proxy features on the katello server indicate the plugin was installed, for example, on the Infrastructure-> Smart proxies- > <proxy devel> ?

(Although the installer ran, and it appears the pulpcore enablement was successful, the smart proxy doesn't show pulpcore as an "Active feature". I don't think this is related to the plugin.)

>    Look at your catalog: sudo podman search `hostname`/
>        This should be empty

Is this issued on the smart proxy for katello server?

>    Find the name of your docker repo from your Katello server. It'll look like > default_organization-environment-content_view-product-image
>    Try a pull that should fail "unauthorized": sudo podman pull `hostname`/<docker repo name>

Also, which host is this issued?

>    Add your docker repo to the unauthenticated repo list via httpie:
> http PUT https://`hostname`/v2/unauthenticated_repository_list repositories:='["<docker repo name>"]'

Same as above, which host?

>    View your catalog again. It should have the repo you added
>    Try to podman pull the docker repo again. It should work now.

Same here, which host?

**Comment by ianballou** [comment]
@jjeffers `hostname` refers to your smart proxy's hostname. All curl queries and podman commands should be directed at your smart proxy.

**Comment by ianballou** [comment]
I've rebased on top of the test infrastructure PR.  The tests are going to fail until I update them.

File: README.md [hunk 1]
```diff
@@ -9,12 +9,42 @@ A Foreman smart proxy plugin for Katello.  Implements container registry functio
...
 
+SSLProxyCheckPeerCN off
+SSLProxyCheckPeerName off
+
 ProxyPass /pulpcore_registry/v2/ http://127.0.0.1:24817/v2/
 ProxyPassReverse /pulpcore_registry/v2/ http://127.0.0.1:24817/v2/
 
-ProxyPass /pulp/container http://127.0.0.1:24816/pulp/container
-ProxyPassReverse http://127.0.0.1:24816/pulp/container  /pulp/container
+ProxyPass /pulp/container/ unix:///run/pulpcore-content.sock|http://centos7-katello-devel.cannolo.example.com/pulp/container/
```
Comment (jjeffers) [inline README.md:18]: This was key!

File: README.md [hunk 2]
```diff
@@ -9,12 +9,42 @@ A Foreman smart proxy plugin for Katello.  Implements container registry functio
...
 ProxyPass /v2 https://127.0.0.1:9090/container_gateway/v2
 ProxyPassReverse https://127.0.0.1:9090/container_gateway/v2 /v2
 ```
+
+# Server configuration
+
+The Container Gateway plugin requires a Pulp 3 instance to connect to.  Related configuration options:
+```
+:pulp_endpoint: 'https://your_pulp_3_server_here.com'
+:pulp_client_ssl_cert: 'Path to X509 certificate for authenticating with Pulp'
```
Comment (jjeffers) [inline README.md:30]: The current foreman-installer doesn't properly set ownership or permissions, so manual adjustments were required to get these keys readable by the plugin.

**Review by jjeffers — APPROVED** [review]
Tested this and after a few updates to compensate for installations that were incomplete, finally able to enable the plugin, add an unauthenticated repo via PUT, then subsequently pull the said container image via the gateway.

**Comment by ianballou** [comment]
Last TODO: figure out how to test the database with GitHub actions.

**Comment by ianballou** [comment]
@jjeffers I've updated the PR with database-related tests.  Lemme know if you want to give it another quick look over.

**Pull request merged**